src/gp_instance.cpp

```diff
@@ -86,3 +130,805 @@ void GPInstance::ProcessLoadedTrees() {
+        pcsp_idx,
+        GetPlvIndex(PlvType::R, dag_nodes_.size(), node->Id()),
+        GetPlvIndex(PlvType::P, dag_nodes_.size(), child_node->Id())
+      });
+    }
+    for (size_t child_idx : node->GetLeafwardRotated()) {
+      auto child_node = dag_nodes_[child_idx];
+      auto pcsp_idx = pcsp_indexer_[node->GetBitset().RotateSubsplit() + child_node->GetBitset()];
+      operations.push_back(Likelihood{
+        pcsp_idx,
+        GetPlvIndex(PlvType::R_TILDE, dag_nodes_.size(), node->Id()),
+        GetPlvIndex(PlvType::P, dag_nodes_.size(), child_node->Id())
+      });
+    }
+  }
+
+  // Compute marginal likelihood.
+  for (size_t root_idx = 0; root_idx < rootsplits_.size(); root_idx++) {
+    auto rootsplit = rootsplits_[root_idx];
+    auto node_idx = subsplit_to_index_[rootsplit + ~rootsplit];
+    operations.push_back(MarginalLikelihood{
+      GetPlvIndex(PlvType::R_HAT, dag_nodes_.size(), node_idx),
+      root_idx,
+      GetPlvIndex(PlvType::P, dag_nodes_.size(), node_idx)
+    });
+  }
+
+  GetEngine()->ProcessOperations(operations);
+}
+
+GPOperationVector GPInstance::MarginalLikelihoodOperations() {
+  // Compute marginal likelihood.
+  GPOperationVector operations;
+  for (size_t i = 0; i < rootsplits_.size(); i++) {
+    auto rootsplit = rootsplits_[i];
+    auto root_subsplit = rootsplit + ~rootsplit;
+    size_t root_idx = subsplit_to_index_[root_subsplit];
+    operations.push_back(MarginalLikelihood{
+      GetPlvIndex(PlvType::R_HAT, dag_nodes_.size(), root_idx),
+      i,
+      GetPlvIndex(PlvType::P, dag_nodes_.size(), root_idx)
+    });
+  }
+  return operations;
+}
+
+void GPInstance::EstimateBranchLengths(double tol, size_t max_iter) {
+  std::cout << "Begin branch optimization\n";
+  GPOperationVector branch_optimization_operations = BranchLengthOptimization();
+  GPOperationVector marginal_lik_operations = MarginalLikelihoodOperations();
+
+  SetRootwardZero();
+  SetLeafwardZero();
+  GetEngine()->ResetLogMarginalLikelihood();
+  PopulatePLVs();
+  ComputeLikelihoods();
+  double current_marginal_log_lik = GetEngine()->GetLogMarginalLikelihood();
+
+  for (size_t i = 0; i < max_iter; i++) {
+    std::cout << "Iteration: " << (i + 1) << std::endl;
+    GetEngine()->ProcessOperations(branch_optimization_operations);
+    GetEngine()->ResetLogMarginalLikelihood();
+    GetEngine()->ProcessOperations(marginal_lik_operations);
+    double marginal_log_lik = GetEngine()->GetLogMarginalLikelihood();
+    std::cout << "Current marginal log likelihood: ";
+    std::cout << std::setprecision(9) << current_marginal_log_lik << std::endl;
+    std::cout << "New marginal log likelihood: ";
+    std::cout << std::setprecision(9) << marginal_log_lik << std::endl;
+    if (marginal_log_lik < current_marginal_log_lik) {
+      std::cout << "Marginal log likelihood decreased. Check branch optimization: adjust step size.\n";
+    }
+    if (abs(current_marginal_log_lik - marginal_log_lik) < tol) {
```

Comment:
Yes, let's turn this into a test!

Reply:
Sorry, I'm not quite sure what you mean here.